In [42]:
import numpy as np
import pandas as pd
from numpy import nan
import math

In [2]:
df = pd.read_csv('data/Restaurant_Scores_-_LIVES_Standard.csv')

In [29]:
#df.head()

## Basic information about the data

In [4]:
print('shape of data: {}'.format(df.shape))

shape of data: (51297, 17)


In [30]:
#df.info()

In [52]:
colnames = df.columns
for colname in colnames:
    count = df[colname].count()
    print('{}: {}'.format(colname, count))

business_id: 51297
business_name: 51297
business_address: 51297
business_city: 51297
business_state: 51297
business_postal_code: 50010
business_latitude: 30296
business_longitude: 30296
business_location: 30296
business_phone_number: 15917
inspection_id: 51297
inspection_date: 51297
inspection_score: 38135
inspection_type: 51297
violation_id: 38728
violation_description: 38728
risk_category: 38728


### Summary of the data
1. The total number of rows is 51297. 
2. Violations id's, violation descriptions, and risk categories are 38728. 
3. 1287 postal codes are missing: 50010
4. Lots of latitudes, longitues, and location are missing: 30296
5. Lots of inspection scores are missing: 38135
6. Violation id's are missing for reinspection/followup, new construction, new ownership,

### Look for inspection types for missing violation id's

In [7]:
df2 = df[['violation_id', 'inspection_type']]

In [8]:
mask = df2['violation_id'].isnull()

In [9]:
df3 = df2[mask]
inspect_types = set()
for atype in df3['inspection_type']:
    inspect_types.add(atype)

In [32]:
#print("Inpection types for missing violation id's: \n{}".format(inspect_types))

## Remove the missing violation id rows. 

In [11]:
# They are not the ones issued with violations.

In [12]:
mask_viol = df['violation_id'].isnull()

In [13]:
mask_viol.shape

(51297,)

In [14]:
df_viol = df[~mask_viol]

In [15]:
df_viol.shape

(38728, 17)

In [33]:
#df_viol.head()

## Clean up the violation id's. 

In [17]:
# 1757_20170928_103131: The first two series of numbers are the inspection id.

In [18]:
df_clean = df_viol[:]

In [19]:
a = df_clean['violation_id']

In [36]:
df_clean.head(20)

business_id                    business_name         business_address  \
0          1757  Dar Bar Pakistani/Indian Cusine             1412 Polk St   
1          4864        DRAGON CITY BAKERY & CAFE          2367 MISSION St   
2         79782                          Deli 23             2449 23rd St   
3         73840         L'acajou Bakery and Cafe       498 09th St Ste. C   
4         76437                  Sweetheart Cafe            909 Grant Ave   
5         82123                Cafe New Honolulu          888 Stockton St   
6          1146              Aux Delices Vietnam             2327 Polk St   
7          3838                      CAFE PICARO             3120 16th St   
8         38929                      Jelly Donut             3198 24th St   
9         81161        Limon Peruvian Rotisserie  1001 South Van Ness Ave   
10        85781                   Domino's #7764             876 Geary St   
11        82880                        Don Pablo             Off The Grid   
14        16441               Hawaiian Drive Inn       2600 San Bruno Ave   
16         2716                     RED A BAKERY           634 CLEMENT St   
17         3165                      HAN IL KWAN           1802 BALBOA St   
18         2995                       CAFE ROCCO           1131 FOLSOM St   
20        64380          Michael Mina Restaurant        252 California St   
21         3594                CELIAS RESTAURANT            4019 JUDAH St   
24         5493   AT&T - (CART 17) DELI [145105]     24 WILLIE MAYS PLAZA   
25        69762                 Chili House Inc.           726 Clement St   

    business_city business_state business_postal_code  business_latitude  \
0   San Francisco             CA                94109          37.789784   
1   San Francisco             CA                94110          37.759174   
2   San Francisco             CA                94110                NaN   
3   San Francisco             CA                94103                NaN   
4   San Francisco             CA                94108                NaN   
5   San Francisco             CA                94108                NaN   
6   San Francisco             CA                94109          37.798274   
7   San Francisco             CA                94103          37.764908   
8   San Francisco             CA                94110          37.752439   
9   San Francisco             CA                94110                NaN   
10  San Francisco             CA                94109                NaN   
11  San Francisco             CA                  NaN                NaN   
14  San Francisco             CA                94134          37.728718   
16  San Francisco             CA                94118          37.782888   
17  San Francisco             CA                94121          37.776599   
18  San Francisco             CA                94103          37.776127   
20  San Francisco             CA                94111          37.793338   
21  San Francisco             CA                94122          37.760423   
24  San Francisco             CA                94107                NaN   
25  San Francisco             CA                94118                NaN   

    business_longitude         business_location  business_phone_number  \
0          -122.420455  (37.789784, -122.420455)                    NaN   
1          -122.419066  (37.759174, -122.419066)           1.415583e+10   
2                  NaN                       NaN                    NaN   
3                  NaN                       NaN                    NaN   
4                  NaN                       NaN                    NaN   
5                  NaN                       NaN                    NaN   
6          -122.422264  (37.798274, -122.422264)                    NaN   
7          -122.422442  (37.764908, -122.422442)           1.415543e+10   
8          -122.416250   (37.752439, -122.41625)                    NaN   
9                  NaN                 

In [21]:
a[0].split('_')[2]

'103131'

In [22]:
L_vid = []
for id in a:
    vid = id.split('_')[2]
    L_vid.append(vid)
ps_vid = pd.Series(L_vid)

In [53]:
# number of violation ids without inspection id in front. this should be 38728
len(ps_vid)

38728

In [55]:
# is there any NaN values?
ps_vid.isnull().any()

False

In [25]:
df_clean['short_violation_id'] = ps_vid

/home/soong/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [56]:
# is there any NaN in df_clean['short_violation_id'] column
df_clean['short_violation_id'].isnull().any()

True

In [60]:
# let's find out why there are NaN values by comparing violation ids and short violation ids
mask_clean = df_clean['short_violation_id'].isnull()
len(mask_clean)

38728

In [63]:
df_clean_diag = df_clean[mask_clean]
df_clean_diag.head()

business_id            business_name           business_address  \
38729        39667      Peet's Coffee & Tea  595 Market St  Suite #143   
38730        91289  Lapats Thai Noodles Bar              601 Larkin St   
38731        89059                 Nande-Ya           1737 Post St 375   
38732        77645              Banana Home              321 Kearny St   
38733        66536            Cypress Grill         99 Harding Park Rd   

       business_city business_state business_postal_code  business_latitude  \
38729  San Francisco             CA                94105          37.789437   
38730  San Francisco             CA                94109                NaN   
38731  San Francisco             CA                94115                NaN   
38732  San Francisco             CA                94108                NaN   
38733  San Francisco             CA                94132          37.723698   

       business_longitude         business_location  business_phone_number  \
38729         -122.401111  (37.789437, -122.401111)                    NaN   
38730                 NaN                       NaN                    NaN   
38731                 NaN                       NaN           1.415369e+10   
38732                 NaN                       NaN                    NaN   
38733         -122.495289  (37.723698, -122.495289)           1.415566e+10   

        inspection_id inspection_date  inspection_score  \
38729  39667_20170912  9/12/2017 0:00              82.0   
38730  91289_20170905   9/5/2017 0:00              89.0   
38731  89059_20170227  2/27/2017 0:00              72.0   
38732  77645_20170320  3/20/2017 0:00              85.0   
38733  66536_20170303   3/3/2017 0:00              90.0   

             inspection_type           violation_id  \
38729  Routine - Unscheduled  39667_20170912_103142   
38730  Routine - Unscheduled  91289_20170905_103144   
38731  Routine - Unscheduled  89059_20170227_103133   
38732  Routine - Unscheduled  77645_20170320_103133   
38733  Routine - Unscheduled  66536_20170303_103149   

                                   violation_description  risk_category  \
38729                   Unclean nonfood contact surfaces       Low Risk   
38730   Unapproved or unmaintained equipment or utensils       Low Risk   
38731             Foods not protected from contamination  Moderate Risk   
38732             Foods not protected from contamination  Moderate Risk   
38733  Wiping cloths not clean or properly stored or ...       Low Risk   

      short_violation_id  
38729                NaN  
38730                NaN  
38731                NaN  
38732                NaN  
38733                NaN

In [68]:
a53 = df_clean_diag['violation_id']

In [74]:
b53 = a53.iloc[0]

In [75]:
b53.split('_')[2]

'103142'

In [26]:
df_clean.head()

business_id                    business_name    business_address  \
0         1757  Dar Bar Pakistani/Indian Cusine        1412 Polk St   
1         4864        DRAGON CITY BAKERY & CAFE     2367 MISSION St   
2        79782                          Deli 23        2449 23rd St   
3        73840         L'acajou Bakery and Cafe  498 09th St Ste. C   
4        76437                  Sweetheart Cafe       909 Grant Ave   

   business_city business_state business_postal_code  business_latitude  \
0  San Francisco             CA                94109          37.789784   
1  San Francisco             CA                94110          37.759174   
2  San Francisco             CA                94110                NaN   
3  San Francisco             CA                94103                NaN   
4  San Francisco             CA                94108                NaN   

   business_longitude         business_location  business_phone_number  \
0         -122.420455  (37.789784, -122.420455)                    NaN   
1         -122.419066  (37.759174, -122.419066)           1.415583e+10   
2                 NaN                       NaN                    NaN   
3                 NaN                       NaN                    NaN   
4                 NaN                       NaN                    NaN   

    inspection_id inspection_date  inspection_score        inspection_type  \
0   1757_20170928  9/28/2017 0:00              86.0  Routine - Unscheduled   
1   4864_20161206  12/6/2016 0:00              84.0  Routine - Unscheduled   
2  79782_20160503   5/3/2016 0:00              92.0  Routine - Unscheduled   
3  73840_20171207  12/7/2017 0:00              71.0  Routine - Unscheduled   
4  76437_20160329  3/29/2016 0:00              76.0  Routine - Unscheduled   

            violation_id                              violation_description  \
0   1757_20170928_103131                   Moderate risk vermin infestation   
1   4864_20161206_103157  Food safety certificate or food handler card n...   
2  79782_20160503_103120             Moderate risk food holding temperature   
3  73840_20171207_103105                           Improper cooling methods   
4  76437_20160329_103113                 Sewage or wastewater contamination   

   risk_category short_violation_id  
0  Moderate Risk             103131  
1       Low Risk             103157  
2  Moderate Risk             103120  
3      High Risk             103105  
4      High Risk             103113

## Creating 7 time periods
y label time period (past 6 months) and features of violations for 6 different time periods:  
0) 11/2017 - 3/2017: y label (0 or 1)  
i) 8/2017 - 10/2017: first period  
ii) 5/2017 - 7/2017: second period  
iii) 2/2017 - 4/2017: third period  
iv) 11/2016 - 1/2017: fourth period  
v) 5/2016 - 10/2016: fifth period  
vi) 3/2015 - 4/2016: sixth period  

In [28]:
df_periods = df_clean[:]

In [44]:
df_periods.to_csv('data/df_periods.csv')

In [48]:
df_periods['short_violation_id'].isnull().sum()

9336